ndb/src/google/cloud/ndb/model.py

```diff
@@ -1833,17 +1833,157 @@ def _db_get_value(self, v, unused_p):
+                :class:`str`.
+            .BadValueError: If the current property is indexed but the UTF-8
+                encoded value exceeds the maximum length (1500 bytes).
+        """
+        if isinstance(value, bytes):
+            try:
+                length = len(value)
+                value = value.decode("utf-8")
+            except UnicodeError:
+                raise exceptions.BadValueError(
+                    "Expected valid UTF-8, got {!r}".format(value)
+                )
+        elif isinstance(value, str):
+            length = len(value.encode("utf-8"))
+        else:
+            raise exceptions.BadValueError(
+                "Expected string, got {!r}".format(value)
+            )
+
+        if self._indexed and length > _MAX_STRING_LENGTH:
+            raise exceptions.BadValueError(
+                "Indexed value %s must be at most %d bytes"
+                % (self._name, _MAX_STRING_LENGTH)
+            )
+
+    def _to_base_type(self, value):
+        """Convert a value to the "base" value type for this property.
+
+        Args:
+            value (Union[bytes, str]): The value to be converted.
+
+        Returns:
+            Optional[bytes]: The converted value. If ``value`` is a
+            :class:`str`, this will return the UTF-8 encoded bytes for it.
+            Otherwise, it will return :data:`None`.
+        """
+        if isinstance(value, str):
+            return value.encode("utf-8")
+
+    def _from_base_type(self, value):
+        """Convert a value from the "base" value type for this property.
+
+        .. note::
+
+            Older versions of ``ndb`` could write non-UTF-8 ``TEXT``
+            properties. This means that if ``value`` is :class:`bytes`, but is
+            not a valid UTF-8 encoded string, it can't (necessarily) be
+            rejected. But, :meth:`_validate` now rejects such values, so it's
+            not possible to write new non-UTF-8 ``TEXT`` properties.
+
+        Args:
+            value (Union[bytes, str]): The value to be converted.
+
+        Returns:
+            Optional[str]: The converted value. If ``value`` is a a valid UTF-8
+            encoded :class:`bytes` string, this will return the decoded
+            :class:`str` corresponding to it. Otherwise, it will return
+            :data:`None`.
+        """
+        if isinstance(value, bytes):
+            try:
+                return value.decode("utf-8")
+            except UnicodeError:
+                pass
+
+    def _db_set_uncompressed_meaning(self, p):
+        """Helper for :meth:`_db_set_value`.
+
+        Raises:
+            NotImplementedError: Always. This method is virtual.
+        """
         raise NotImplementedError
 
 
 class StringProperty(TextProperty):
+    """An indexed property that contains UTF-8 encoded text values.
+
+    This is nearly identical to :class:`TextProperty`, but is indexed. Values
+    must be at most 1500 bytes (when encoded from UTF-8 to bytes).
```

Comment:
Encoding *to* UTF-8, not from.

Reply:
Yeah I didn't like that sentence when I wrote it. I really meant from `str`, so I'll write that.